pennylane/transforms/rz_phase_gradient.py

```diff
@@ -0,0 +1,215 @@
+    r"""Quantum function transform to decompose all instances of :class:`~.RZ` gates into additions
+    using a phase gradient resource state.
+
+    For example, an :class:`~.RZ` gate with angle :math:`\phi = (0 \cdot 2^{-1} + 1 \cdot 2^{-2} + 0 \cdot 2^{-3}) 2\pi`
+    is translated into the following routine, where the angle is conditionally prepared on the ``angle_wires`` in binary
+    and added to a ``phase_grad_wires`` register semi-inplace via :class:`~.SemiAdder`.
+
+    .. code-block::
+
+        target: ─RZ(ϕ)─ = ────╭●──────────────╭●────exp(iϕ/2)─┤
+         ang_0:           ────├|0⟩─╭SemiAdder─├|0⟩────────────┤
+         ang_1:           ────├|1⟩─├SemiAdder─├|1⟩────────────┤
+         ang_2:           ────╰|0⟩─├SemiAdder─╰|0⟩────────────┤
+         phg_0:           ─────────├SemiAdder─────────────────┤
+         phg_1:           ─────────├SemiAdder─────────────────┤
+         phg_2:           ─────────╰SemiAdder─────────────────┤
+
+    For this routine to work, the provided ``phase_grad_wires`` need to hold a phase gradient
+    state :math:`|\nabla Z\rangle = \frac{1}{\sqrt{2^n}} \sum_{m=0}^{2^n-1} e^{2 \pi i \frac{m}{2^n}} |m\rangle`.
+    Because this state is not modified and can be re-used at a later stage, the transform does not prepare it but
+    rather assumes it has been prepared on those wires at an earlier stage.
+
+
+    Note that :class:`~.SemiAdder` requires additional ``work_wires`` (not shown in the diagram) for the semi-in-place addition
+    :math:`\text{SemiAdder}|x\rangle_\text{ang} |y\rangle_\text{phg} = |x\rangle_\text{ang} |x + y\rangle_\text{phg}`.
+
+    More details can be found on page 4 in `arXiv:1709.06648 <https://arxiv.org/abs/1709.06648>`__
+    and Figure 17a in `arXiv:2211.15465 <https://arxiv.org/abs/2211.15465>`__ (a generalization to
+    multiplexed :class:`~.RZ` rotations is provided in Figure 4 in
+    `arXiv:2409.07332 <https://arxiv.org/abs/2409.07332>`__).
+
+    Note that technically, this circuit realizes :class:`~.PhaseShift`, i.e. :math:`R_\phi(\phi) = R_Z(\phi) e^{i\phi/2}`.
+    The additional global phase is taken into account in the decomposition.
+
+    Args:
+        tape (QNode or QuantumTape or Callable): A quantum circuit containing :class:`~.RZ` gates.
+        angle_wires (Wires): The qubits that conditionally load the angle :math:`\phi` of
+            the :class:`~.RZ` gate in binary as a multiple of :math:`2\pi`.
+            The length of the ``angle_wires`` implicitly determines the precision
+            with which the angle is represented.
+            E.g., :math:`(2^{-1} + 2^{-2} + 2^{-3}) 2\pi` is exactly represented by three bits as ``111``.
+        phase_grad_wires (Wires): The catalyst qubits with a phase gradient state prepared on them.
+            Needs to be at least the length of ``angle_wires`` and will only
+            use the first ``len(angle_wires)`` according to the precision with which the angle is decomposed.
```

Comment:
> according to the precision with which the angle is decomposed.

Not sure what this part means. Does it add anything beyond saying we will only use the first ``len(angle_wires)`` wires?
Is there a situation where we use more than `len(angle_wires)`? This may have been discussed elsewhere :thinking: feel free to resolve if it has.